apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -208,30 +208,38 @@ defmodule Aecore.Peers.Sync do
   end
 
   # Gets all unknown blocks, starting from the given one
-  defp check_peer_block(peer_uri, block_hash, state) do
-    case Chain.has_block?(block_hash) do
-      false ->
-        case(HttpClient.get_block({peer_uri, block_hash})) do
-          {:ok, deserialized_block} ->
+  defp check_peer_blocks(peer_uri, from_block_hash, state) do
+    case HttpClient.get_raw_blocks({peer_uri, from_block_hash, Chain.top_block_hash()}) do
+      {:ok, blocks} ->
+        if(Enum.count(blocks) > 0) do
+          state = Enum.reduce(blocks, state, fn(block, state_acc) ->
             try do
-              BlockValidation.single_validate_block(deserialized_block)
-              peer_block_hash =
-                BlockValidation.block_header_hash(deserialized_block.header)
-
-              if(block_hash == peer_block_hash) do
-                check_peer_block(peer_uri, deserialized_block.header.prev_hash,
-                  Map.put(state, peer_block_hash, deserialized_block))
-              else
-                state
+              BlockValidation.single_validate_block(block)
+              peer_block_hash = BlockValidation.block_header_hash(block.header)
+
+              case Chain.has_block?(peer_block_hash) do
+                false ->
+                  Map.put(state_acc, peer_block_hash, block)
+                true ->
+                  state_acc
               end
             catch
               {:error, _} ->
-                state
+                state_acc
             end
-          :error ->
+          end)
+
+          earliest_block = Enum.at(blocks, Enum.count(blocks) - 1)
+          earliest_block_hash = BlockValidation.block_header_hash(earliest_block.header)
+          if(earliest_block_hash != Chain.top_block_hash()) do
```

Comment:
What if we have `earliest_block_hash`, but it's not the top_block, because a fork happened?
You should use `Chain.has_block?` to do such checks.